swiftsimio/visualisation.py

```diff
@@ -0,0 +1,195 @@
+"""
+Contains visualisation functions that are accelerated with numba
+if it is available.
+"""
+
+from typing import Union
+from math import sqrt
+from numpy import float32, int32, zeros, array, arange, ndarray, ones
+from swiftsimio import SWIFTDataset
+
+from swiftsimio.accelerated import jit
+
+kernel_gamma = 1.778002
+kernel_constant = 80.0 * 3.14159 / 7.0
+
+
+@jit(nopython=True, fastmath=True)
+def kernel(r: Union[float, float32], h: Union[float, float32]):
+    """
+    Kernel implementation for swiftsimio. This is the Wendland-C2
+    kernel as shown in Denhen & Aly (2012).
+
+    Give it a radius and a smoothing length, and it returns the
+    contribution to the density.
+    """
+    inverse_H = 1.0 / (h * kernel_gamma)
+    ratio = r * inverse_H
+
+    kernel = 0.0
+
+    if ratio < 1.0:
+        ratio_2 = ratio * ratio
+        ratio_3 = ratio_2 * ratio
+
+        if ratio < 0.5:
+            kernel += 3.0 * ratio_3 - 3.0 * ratio_2 + 0.5
+
+        else:
+            kernel += -1.0 * ratio_3 + 3.0 * ratio_2 - 3.0 * ratio + 1.0
+
+        kernel *= kernel_constant * inverse_H * inverse_H
+
+    return kernel
+
+
+@jit(nopython=True)
+def pair_max(a: int32, b: int32) -> int32:
+    if a > b:
+        return a
+    else:
+        return b
+
+
+@jit(nopython=True)
+def pair_min(a: int32, b: int32) -> int32:
+    if a < b:
+        return a
+    else:
+        return b
+
+
+@jit(nopython=True, fastmath=True)
+def scatter(x: float32, y: float32, m: float32, h: float32, res: int) -> ndarray:
```

Comment:
I presume the rest of swiftsimio is float32?

Reply:
That's a good point. This should actually be `float64` as we store positions as doubles.